execution/evm/contract.go

```diff
@@ -862,6 +862,22 @@ func (c *Contract) jump(code []byte, to uint64, pc *uint64) error {
 	return nil
 }
 
+// isInsidePushData checks if the operator code(n) is inside push data
+func isInsidePushData(code []byte, n uint64) bool {
+	if uint64(len(code)) <= n {
+		return false
+	}
+	i := uint64(0)
+	for i < n {
```

Comment:
This could get expensive for long code. A simple improvement on the average case would be to scan from `n` backwards to the first `PUSH<N>` opcode and see whether n falls within [i, i+N]. The worst case (no pushes) is the same, but there are usually plenty of pushes.

I note go-ethereum amortises this work by creating a bitmap that partitions each symbol in the code into code/not code (https://github.com/ethereum/go-ethereum/blob/master/core/vm/analysis.go). I think I prefer the simple scanning approach. I suspect there are enough pushes in real solidity code that the 'count back from n' algorithm is good enough in practice.

Another ahead-of-time approach would be to store an array of offsets for all `PUSH`es in the code, then perform a binary search to find the nearest PUSH to the left.

What do people think about this? Should we bite the bullet and eat the cost for this at contract deploy time. If so which of the two approaches above seems best? If we are doing a linear pass of the code then we also have an opportunity to capture other metadata, not sure if there is anything that might be useful there.

Reply:
I agree this is expensive, but I don't think the 'count back from n' algorithm would work.
For example, in `PUSH1, PUSH1, JUMPDEST` the second `PUSH1` is data rather than opcode. If we simply scan backwards `JUMPDEST` would be treated as data as the second `PUSH1`.
I think the other algorithm you mentioned that stores an array of offsets for all `PUSH`es is a good one.

Reply:
Yes you are right!

However we can:

Define `P(n)` to be the index of the closest previous push that captures `n`, that is:

`P(n)` defined as the largest integer such that `P(n) < n` and `code[P(n)] = PUSH<N>` where `n-P(n) <= N` or else null.

Then for a `JUMPDEST` at `n` we can search for`P(n)` then ask 'is this a real push?' by searching for `P(P(n))`. If `P(P(n))` is null then we are done and `P(n)` must be real, and therefore `JUMPDEST` is just data.

However if `P(P(n))` exists then we must ask i's `P(P(n))` a real push'. We can carry on searching backwards recursively like this. It will terminate when we there is no previous capturing push. In a chain like this the leftmost push is the first 'real one', the odd-numbered ones are just data so eventually you need to check if the rightmost real push captures n (since P(n) may be data, but P(P(n)) might still capture n).

So having thought that through it seems like quite a bit of work for a still sub-optimal solution!

Reply:
Cool! I'll implement this.

Reply:
Oops, see my comment below - I do not think this is worth doing - I think we should do the bitmap.